R/tm_rmarkdown.R

```diff
@@ -0,0 +1,267 @@
+  args <- as.list(environment())
+
+  ans <- module(
+    label = label,
+    server = srv_rmarkdown,
+    server_args = list(rmd_file = rmd_file, allow_download = allow_download, extra_transform = extra_transform),
+    ui = ui_rmarkdown,
+    ui_args = args,
+    transformators = transformators,
+    datanames = datanames
+  )
+  # attr(ans, "teal_bookmarkable") <- TRUE
+  ans
+}
+
+# UI function for the rmarkdown module
+ui_rmarkdown <- function(id, rmd_file, allow_download, extra_transform, ...) {
+  args <- list(...)
+  ns <- NS(id)
+
+  teal.widgets::standard_layout(
+    output = teal.widgets::white_small_well(
+      tags$div(
+        tags$h4(
+          "Rendered report from: ",
+          tags$code(basename(rmd_file))
+        ),
+        if (allow_download) {
+          downloadButton(
+            ns("download_rmd"),
+            sprintf("Download '%s'", basename(rmd_file)),
+            class = "btn-primary btn-sm"
+          )
+        },
+        ui_transform_teal_data(ns("extra_transform"), transformators = extra_transform)
+      ),
+      tags$hr(),
+      uiOutput(ns("rmd_output"))
+    ),
+    encoding = NULL,
+    pre_output = args$pre_output,
+    post_output = args$post_output
+  )
+}
+
+# Server function for the rmarkdown module
+srv_rmarkdown <- function(id, data, rmd_file, allow_download, extra_transform) {
+  checkmate::assert_class(data, "reactive")
+  checkmate::assert_class(isolate(data()), "teal_data")
+  moduleServer(id, function(input, output, session) {
+    if (allow_download) {
+      output$download_rmd <- downloadHandler(
+        filename = function() basename(rmd_file),
+        content = function(file) file.copy(rmd_file, file),
+        contentType = "text/plain"
+      )
+    }
+
+    pre_decorated_q_r <- reactive({
+      data_q <- req(data())
+      teal.reporter::teal_card(data_q) <- c(
+        teal.reporter::teal_card(data_q),
+        teal.reporter::teal_card("## Module's output(s)")
+      )
+      eval_code(
+        data_q,
+        sprintf(
+          "rmd_data <- list(%s)",
+          toString(sprintf("%1$s = %1$s", sapply(names(data_q), as.name)))
+        )
+      )
+    })
+
+    q_r <- data_with_output_decorated <- teal::srv_transform_teal_data(
+      "extra_transform",
+      data = pre_decorated_q_r,
+      transformators = extra_transform
+    )
+
+    rendered_path_r <- reactive({
+      datasets <- req(q_r()) # Ensure data is available
+      temp_dir <- tempdir()
+      temp_rmd <- tempfile(tmpdir = temp_dir, fileext = ".Rmd")
+      temp_html <- tempfile(tmpdir = temp_dir, fileext = ".md")
+      file.copy(rmd_file, temp_rmd) # Use a copy of the Rmd file to avoid modifying the original
+
+      tryCatch(
+        {
+          rmarkdown::render(
+            temp_rmd,
+            output_format = rmarkdown::md_document(
+              variant = "gfm",
+              toc = TRUE,
+              preserve_yaml = TRUE
+            ),
+            output_file = temp_html,
```

Comment:
Probably `output_file` should remain empty as it will be given automatically - depending on the `output:` in yaml header. It could be `md/html/pdf` etc.
```suggestion
```

Reply:
Good point, `render()` returns the path